x-pack/plugin/repository-encrypted/src/main/java/org/elasticsearch/repositories/encrypted/EncryptedRepository.java

```diff
@@ -0,0 +1,307 @@
+        }
+
+        @Override
+        public InputStream readBlob(String blobName) throws IOException {
+            final BytesReference dataDecryptionKeyBytes = Streams.readFully(this.encryptionMetadataBlobContainer.readBlob(blobName));
+            final SecretKey dataDecryptionKey;
+            try {
+                dataDecryptionKey = unwrapKey(BytesReference.toBytes(dataDecryptionKeyBytes), this.masterSecretKey);
+            } catch (NoSuchAlgorithmException | NoSuchPaddingException | InvalidKeyException e) {
+                throw new IOException(e);
+            }
+            return GCMPacketsCipherInputStream.getDecryptor(this.delegatedBlobContainer.readBlob(blobName), dataDecryptionKey, NONCE,
+                    BC_FIPS_PROVIDER);
+        }
+
+        @Override
+        public void writeBlob(String blobName, InputStream inputStream, long blobSize, boolean failIfAlreadyExists) throws IOException {
+            final byte[] wrappedDataEncryptionKey;
+            final SecretKey dataEncryptionKey;
+            try {
+                dataEncryptionKey = generateRandomSecretKey();
+                wrappedDataEncryptionKey = wrapKey(dataEncryptionKey, this.masterSecretKey);
+            } catch (NoSuchAlgorithmException | NoSuchPaddingException | InvalidKeyException | IllegalBlockSizeException e) {
+                throw new IOException(e);
+            }
+            try (InputStream stream = new ByteArrayInputStream(wrappedDataEncryptionKey)) {
+                this.encryptionMetadataBlobContainer.writeBlob(blobName, stream, wrappedDataEncryptionKey.length, failIfAlreadyExists);
+            }
+            final long encryptedSize = GCMPacketsCipherInputStream.getEncryptionSizeFromPlainSize(blobSize);
+            try (InputStream encryptedInputStream = GCMPacketsCipherInputStream.getEncryptor(inputStream, dataEncryptionKey, NONCE,
+                    BC_FIPS_PROVIDER)) {
+                this.delegatedBlobContainer.writeBlob(blobName, encryptedInputStream, encryptedSize, failIfAlreadyExists);
+            }
+        }
+
+        @Override
+        public void writeBlobAtomic(String blobName, InputStream inputStream, long blobSize, boolean failIfAlreadyExists)
+                throws IOException {
+            // does not support atomic write
+            writeBlob(blobName, inputStream, blobSize, failIfAlreadyExists);
+        }
+
+        @Override
+        public void deleteBlob(String blobName) throws IOException {
+            this.delegatedBlobContainer.deleteBlob(blobName);
+            this.encryptionMetadataBlobContainer.deleteBlob(blobName);
+        }
+
+        @Override
+        public DeleteResult delete() throws IOException {
+            DeleteResult result = this.delegatedBlobContainer.delete();
+            this.encryptionMetadataBlobContainer.delete();
+            return result;
+        }
+
+        @Override
+        public Map<String, BlobMetaData> listBlobs() throws IOException {
```

Comment:
I think we should base this method and  `listBlobsByPrefix` on the metadata instead of the data. We are using listing for two purposes:
1. Finding the latest index-N blob (that will change but for now it is what it is and should work correctly)
2. Finding unreferenced blobs before cleaning them up

The former will only work correctly for blobs that have their metadata in-tact. If we delete the encryption metadata first, we must not list out dangling data blobs for which the metadata is gone imo. Otherwise we fundamentally change the behaviour of the blob container and the guarantees we're assuming for its operations.

The fact that this will break the cleanup should be addressed by overriding the cleanup logic accordingly (see my other comment on that).

Reply:
From https://github.com/elastic/elasticsearch/pull/48221#discussion_r337218033 I see that what you expect is that _if the metadata exists **then** the encrypted data exists as well_.

But I wrote it the other way around: _if the encrypted data exists **then** the metadata exists as well_.

This is the perspective from which _write_, _delete_ and _listBlobsByPrefix_ are written, **please correct me if I got it wrong**. For the _read_ operation, the relation is not important because both must be available for the operation to succeed.

I have chosen this relation and not the obverse you suggested because encrypted data without the decryption key is a scary thing. But also, practically, it is much more efficient to leak encryption metadata than the actual data, given the size discrepancy between the two (and my choosing tolerates orphaned encryption metadata).